### PR TITLE
requireCamelCaseOrUpperCaseIdentifiers: add exception options

### PR DESCRIPTION
**Pull request by pabigot**
>Some coding guidelines might want optional parameters to start with `opt_` while still checking conformance of the remainder of the name, or have specific identifiers like `var_args` that are exceptions.
>
>Other projects may want to use this rule but require that values that hold measurements carry a suffix specifying the units of measurement; e.g. `ambientTemperature_dCel` or `centerFrequency_kHz`.  The interpretation of units can be misleading when conjoined with the identifier without a separating underscore.
>
>This commit adds options `allExcept`, `allowedPrefixes`, and `allowedSuffixes`.  All three recognize an array of strings, RegExps, or ESTree RegExpLiteral objects that identify strings or string sequences. Identifiers that match strings/patterns in `allExcept` are excluded from validation by the rule.  Identifiers that start (end) with a match from `allowedPrefixes` (`allowedSuffixes`) are checked after the match is removed from the identifier.

Resubmission following comments in the original pull request #2028 and feature issue #2029.  The primary changes are:

* Add [ESTree RegExpLiteral](https://github.com/estree/estree/blob/master/spec.md#regexpliteral) objects so regular expressions can be expressed in JSON config specs;
* Incorporate the prefix/suffix separator character (generally but not always `_`) into the strings and patterns;
* Add `allExcept` to complete the functionality required to close issue #1900.

### CODE REVIEW

File: lib/rules/require-camelcase-or-uppercase-identifiers.js [hunk 1]
```diff
@@ -79,10 +88,161 @@
...
+            try {
+                // ESTree RegExpLiteral ok if it produces RegExp
+                rv.push(new RegExp(elt.regex.pattern, elt.regex.flags || ''));
+            } catch (e) {
+                // Not a valid RegExpLiteral
+                rv = null;
+            }
+        } else {
+            // Unknown value
+            rv = null;
```
Comment (gibson042) [inline lib/rules/require-camelcase-or-uppercase-identifiers.js]: An `assert` (and corresponding exception) is probably better than these `rv = null` silent failures.

Comment (gibson042) [inline lib/rules/require-camelcase-or-uppercase-identifiers.js]: Nevermind, I see that it's just handled elsewhere.

**Comment by gibson042** [comment]
I like this a lot.

**Comment by markelog** [comment]
@pabigot Nicely done! note @gibson042 comment, that is the high praise :)

**Comment by pabigot** [comment]
Thanks; I like it too.

Please let me know if you have additional changes you'd like made, and if you intend to merge the amend commit with the main one or would prefer I do it.

File: lib/rules/require-camelcase-or-uppercase-identifiers.js [hunk 2]
```diff
@@ -79,10 +88,161 @@
  * var snake_case = { SnakeCase: 6 };
  * ```
  *
+ * ##### Valid for `{ allowedPrefix: ["opt_", /pfx\d+_/] }`
```
Comment (markelog) [inline lib/rules/require-camelcase-or-uppercase-identifiers.js]: I think we need how we could store regexp in json

Comment (markelog) [inline lib/rules/require-camelcase-or-uppercase-identifiers.js]: Oh, i see, you already did that :)

**Pull request closed without merging**